tests/integrationv2/test_early_data.py

```diff
@@ -0,0 +1,397 @@
+    early_data_file = str(tmp_path / EARLY_DATA_FILE)
+    early_data = get_early_data_bytes(early_data_file, early_data_size)
+
+    options = ProviderOptions(
+        port=next(available_ports),
+        cipher=cipher,
+        protocol=protocol,
+        insecure=True,
+        use_session_ticket=True,
+        reconnect=True,
+    )
+    options.ticket_file = None
+    options.early_data_file = early_data_file
+    options.max_early_data = MAX_EARLY_DATA
+
+    client_options = copy.copy(options)
+    client_options.mode = Provider.ClientMode
+
+    server_options = copy.copy(options)
+    server_options.mode = Provider.ServerMode
+    server_options.key = certificate.key # Required for the initial connection
+    server_options.cert = certificate.cert # Required for the initial connection
+    server_options.reconnects_before_exit = NUM_CONNECTIONS
+
+    server = managed_process(provider, server_options)
+    s2n_client = managed_process(S2N, client_options)
+
+    for results in s2n_client.get_results():
+        assert results.exception is None
+        assert results.exit_code == 0
+        assert results.stdout.count(to_bytes(S2N_EARLY_DATA_ACCEPTED_MARKER)) == NUM_RESUMES
+
+    for results in server.get_results():
+        assert results.exception is None
+        assert results.exit_code == 0
+        assert results.stdout.count(early_data) == NUM_RESUMES
+
+
+"""
+Test the S2N server rejecting early data.
+
+We do this by disabling early data on the server after the ticket is issued.
+When the client attempts to use the ticket to send early data, the server rejects the attempt.
+
+We can't perform an S2N client version of this test because the S2N client performs its hardcoded
+reconnects automatically, without any mechanism to modify the connection in between.
+"""
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", TLS13_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("curve", ALL_TEST_CURVES, ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", ALL_TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", [Protocols.TLS13], ids=get_parameter_name)
+@pytest.mark.parametrize("provider", CLIENT_PROVIDERS, ids=get_parameter_name)
+@pytest.mark.parametrize("early_data_size", [int(MAX_EARLY_DATA/2), int(MAX_EARLY_DATA-1), MAX_EARLY_DATA, 1])
+def test_s2n_server_with_early_data_rejected(managed_process, tmp_path, cipher, curve, protocol, provider, certificate, early_data_size):
+    ticket_file = str(tmp_path / TICKET_FILE)
+    early_data_file = str(tmp_path / EARLY_DATA_FILE)
+    early_data = get_early_data_bytes(early_data_file, early_data_size)
+
+    options = ProviderOptions(
+        port=next(available_ports),
+        cipher=cipher,
+        curve=curve,
+        protocol=protocol,
+        insecure=True,
+        use_session_ticket=True,
+        data_to_send=DATA_TO_SEND,
```

Comment:
Same as above: 

We should be setting this parameter only on the client options and not server, so we can move below
`client_options.data_to_send = DATA_TO_SEND`